### PR TITLE
Extend library models to mark fields as nullable

### PR DESCRIPTION
**Pull request by nimakarimipour**
This PR extends `LibraryModels` to add support for marking fields `@Nullable`. This feature is required to enable [Annotator](https://github.com/ucr-riple/NullAwayAnnotator) index impact of making fields `@Nullable` on downstream dependencies.

Please note, since this feature is mostly going to be used while indexing impacts on downstream dependencies, this PR does not focus on the optimality of the implementation. Also the working set in iterations is not expected to be large. We can optimize the implementation in a follow up PR if needed.

### CODE REVIEW

File: nullaway/src/main/java/com/uber/nullaway/LibraryModels.java
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/LibraryModels.java:257]: Can we just use AutoValue for this class, since we already rely on it?  Would get rid of a lot of boilerplate.  See, e.g.:

https://github.com/uber/NullAway/blob/master/nullaway/src/main/java/com/uber/nullaway/dataflow/DataFlow.java#L318-L318

File: nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java:261]: Let's add a bailout check here: if `libraryModels.nullableFields().isEmpty()`, immediately return `false`.  That should mitigate any performance impact for current use cases.

File: test-java-lib/src/main/java/com/uber/lib/unannotated/UnannotatedWithModels.java
Comment (msridhar) [inline test-java-lib/src/main/java/com/uber/lib/unannotated/UnannotatedWithModels.java:7]: What is the purpose of having two fields here?  Are we testing something different with each field?

File: nullaway/src/test/java/com/uber/nullaway/NullAwayCustomLibraryModelsTests.java
Comment (msridhar) [inline nullaway/src/test/java/com/uber/nullaway/NullAwayCustomLibraryModelsTests.java:251]: Can we also test that dereferencing a library field modeled as `@Nullable` causes NullAway to report an error?

**Review by msridhar — COMMENTED** [review]
Overall looks really good!  A few comments

File: nullaway/src/main/java/com/uber/nullaway/LibraryModels.java
Comment (nimakarimipour) [reply nullaway/src/main/java/com/uber/nullaway/LibraryModels.java]: @msridhar Sure. Used `@AutoValue` : f15b983c1b68079923bc25477ddf7c1df515d237

File: test-java-lib/src/main/java/com/uber/lib/unannotated/UnannotatedWithModels.java
Comment (nimakarimipour) [reply test-java-lib/src/main/java/com/uber/lib/unannotated/UnannotatedWithModels.java:7]: Added in 819f75e48fa87ef50eae02ec008fc19fbdfbcadf

File: nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java
Comment (nimakarimipour) [reply nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java:265]: Added in 819f75e48fa87ef50eae02ec008fc19fbdfbcadf

File: nullaway/src/test/java/com/uber/nullaway/NullAwayCustomLibraryModelsTests.java
Comment (nimakarimipour) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayCustomLibraryModelsTests.java:251]: Sure, I thought it might be a good idea to add a fresh method in the same test for dereference:
0f73d506c657bea941a9a22147e3049af1ee966b

**Comment by nimakarimipour** [comment]
@msridhar Thanks for the review, this is ready for another round. I think it might be better to not land this even after approvals, and land it once I can locally test it with the updates on Annotator.

File: test-java-lib/src/main/java/com/uber/lib/unannotated/UnannotatedWithModels.java
Comment (msridhar) [reply test-java-lib/src/main/java/com/uber/lib/unannotated/UnannotatedWithModels.java:7]: Doesn't hurt anything I guess, we can leave it :-)

File: nullaway/src/test/java/com/uber/nullaway/NullAwayCustomLibraryModelsTests.java
Comment (msridhar) [inline nullaway/src/test/java/com/uber/nullaway/NullAwayCustomLibraryModelsTests.java:232]: I realized we need to support / test one more scenario.  If code writes a `@Nullable` value to a library field modeled as `@Nullable` there should be no error.  Not sure that will work right now.  Can we test?

**Review by msridhar — COMMENTED** [review]
one more thing

File: nullaway/src/test/java/com/uber/nullaway/NullAwayCustomLibraryModelsTests.java
Comment (nimakarimipour) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayCustomLibraryModelsTests.java:232]: @msridhar The test fails. The impact of not having this support is that Annotator under rare circumstances might reject a fix tree that there is a direct nullable assignment in downstream dependencies to one of the fields in the tree which also the overall impact is less than 0. Please let me know if we want this support or you think we should just skip for now. Overall, I think for consistency we should have this support.

One way to fix that is that we need to add a custom handling for checking if the field is annotated as `@Nullable` [here](https://github.com/uber/NullAway/blob/5fbee1ff8682762cd2f291070491fe3e476787cf/nullaway/src/main/java/com/uber/nullaway/NullAway.java#L489). That might require adding a new method to handlers (e.g. `hasNullableAnnotation`, please let me know if we can use existing methods). In that case please let me know if I should add this support in this PR or have a followup PR after this one lands. Thank you.

Comment (msridhar) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayCustomLibraryModelsTests.java:232]: I agree that it is likely rare that we would want to model a library field as `@Nullable` _and_ clients of the library are writing `@Nullable` values into the field.  But I worry that the incomplete support will bite us down the road, and even that we'll run into some unexpected case with the annotator.  So I say let's just implement the support in this PR.

I agree with your strategy of adding a new `Handler` method; we can declare it `boolean onOverrideFieldNullability(Symbol fieldSymbol)` to keep things simple for now.  It only gets called if the field is not annotated `@Nullable`, and if it returns `true`, the field should be treated as `@Nullable` anyway.  Let me know if it doesn't make sense.  Thanks!

Comment (nimakarimipour) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayCustomLibraryModelsTests.java:232]: @msridhar Sure, sounds good. But I also found the method below in handlers API:
`Nullness onOverrideMethodReturnNullability(Symbol.MethodSymbol methodSymbol,
      VisitorState state,
      boolean isAnnotated,
      Nullness returnNullness)`

I thought it might be more consistent with other methods and its name (overriding the passed value as arg), if we have the method declared as:
`Nullness onOverrideFieldNullability(Symbol field, Nullness fieldNullness);`

We can also comply with the contract you mentioned for calling the method in the method body. Please let me know if I should use a different declaration. Thank you.

8347141ce8ddc3323b3cc5231759d881031b5c12

Comment (msridhar) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayCustomLibraryModelsTests.java:232]: I suggested the contract above for performance and to limit unnecessary generality.  With method return nullability I think we need more complex overriding functionality, which is why it has the rather complex behavior and signature.  With fields, I personally think it's ok to just say if any handler makes a field nullable, then it's nullable.  What do you think?  I personally prefer the simpler approach and then refactoring later if we need it.

Comment (nimakarimipour) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayCustomLibraryModelsTests.java:232]: @msridhar Sure, makes sense. We really don't need flexibility here. Please find the updated version 4d8187131ec57e117903d302a8d7d096b21b250b

Comment (msridhar) [reply nullaway/src/test/java/com/uber/nullaway/NullAwayCustomLibraryModelsTests.java:232]: Thanks this looks great now.

File: nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java
Comment (msridhar) [inline nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java:286]: One last question.  Why not just do `return libraryModels.nullableFields().contains(fieldRef(enclosingClassName,fieldName))`?  We don't care about performance too much, but that should be constant time assuming hashing, and also it seems clearer.

**Review by msridhar — COMMENTED** [review]
One more thing

File: nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java
Comment (nimakarimipour) [reply nullaway/src/main/java/com/uber/nullaway/handlers/LibraryModelsHandler.java]: Sure, make sense. 1728198fabe668b5f93929efbf1b5d517d6f3a3b

**Review by msridhar — APPROVED** [review]
LGTM, thanks!!

**Pull request merged**
